docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md

```diff
@@ -0,0 +1,330 @@
+type Handler struct {
+	cli cloudevents.Client
+}
+
+// NewHandler returns a new Handler for the given CloudEvents client.
+func NewHandler(c cloudevents.Client) *Handler {
+	rand.Seed(time.Now().UnixNano())
+
+	return &Handler{
+		cli: c,
+	}
+}
+
+// Run starts the handler and blocks until it returns.
+func (h *Handler) Run(ctx context.Context) error {
+	return h.cli.StartReceiver(ctx, h.receive)
+}
+
+// ACKResponse represents the data of a CloudEvent payload returned to
+// acknowledge the processing of an event.
+type ACKResponse struct {
+	Code   ACKCode     `json:"code"`
+	Detail interface{} `json:"detail"`
+}
+
+// ACKCode defines the outcome of the processing of an event.
+type ACKCode int
+
+// Enum of supported ACK codes.
+const (
+	CodeSuccess ACKCode = iota
+	CodeFailure
+)
+
+// receive implements the handler's receive logic.
+func (h *Handler) receive(e cloudevents.Event) (*cloudevents.Event, cloudevents.Result) {
+	code := CodeSuccess
+	ceResult := cloudevents.ResultACK
+
+	result, err := processEvent(e)
+	// this error handling is for demo purposes only, since processEvent()
+	// always succeeds
+	if err != nil {
+		code = CodeFailure
+		ceResult = cloudevents.ResultNACK
+	}
+
+	return newAckEvent(e, code, result), ceResult
+}
+
+// processEvent processes the event and returns the result of the processing.
+func processEvent(e cloudevents.Event) (interface{} /*result*/, error) {
+	tBegin := time.Now()
+
+	// simulate processing time
+	randomDelay()
+
+	res := &exampleResult{
+		Message:        "event processed successfully",
+		ProcessingTime: time.Since(tBegin).Milliseconds(),
+	}
+
+	return res, nil
+}
+
+// exampleResult represents a fictional structured result of some event
+// processing.
+type exampleResult struct {
+	Message        string `json:"message"`
+	ProcessingTime int64  `json:"processing_time_ms"`
+}
+
+// newAckEvent returns a CloudEvent that acknowledges the processing of an
+// event.
+func newAckEvent(e cloudevents.Event, code ACKCode, detail interface{}) *cloudevents.Event {
+	resp := cloudevents.NewEvent()
+	resp.SetType(eventTypeAck)
+	resp.SetSource(eventSrcName)
+	resp.SetExtension(ceExtProcessedType, e.Type())
+
+	data := &ACKResponse{
+		Code:   code,
+		Detail: detail,
+	}
+
+	if err := resp.SetData(cloudevents.ApplicationJSON, data); err != nil {
+		log.Panicf("Error serializing CloudEvent data: %s", err)
+	}
+
+	return &resp
+}
+
+// randomDelay sleeps for a random amount of time.
+func randomDelay() {
+	const maxDelay = 100 * time.Millisecond
+
+	randomDelay := time.Duration(rand.Int63n(int64(maxDelay)))
+	time.Sleep(randomDelay)
+}
+
+```
+
+Create the `go.mod` file and run `tidy`:
+```
+go mod init
+go mod tidy
+```
+
+Create a new file `Dockerfile` in the same folder:
+```dockerfile
+FROM golang:1.15-buster AS builder
+ENV CGO_ENABLED 0
+WORKDIR /project
+COPY . ./
+RUN go build
+FROM gcr.io/distroless/static:nonroot
+COPY --from=builder /project/demosink /
+ENTRYPOINT ["/demosink"]
+```
```

Comment:
```suggestion
1. Create a new file `Dockerfile` in the same folder:
    
    ```dockerfile
    FROM golang:1.15-buster AS builder
    ENV CGO_ENABLED 0
    WORKDIR /project
    COPY . ./
    RUN go build
    FROM gcr.io/distroless/static:nonroot
    COPY --from=builder /project/demosink /
    ENTRYPOINT ["/demosink"]
    ```
```

suggestion of what I meant wrt. numbering and indentation 🙂

Reply:
repeat for all below steps...let me know if you have any trouble; I can just add a clean up PR after this merges if you do.